### DISCUSSION TITLE
Beta Channel Setup

### DISCUSSION THREAD

**Discussion by davidnorris78** [sparkle-project/Sparkle#2189] [General]
I'm trying to get a Beta Channel setup in a number of our apps.  I've updated them to the latest version of Sparkle.  

I understand about updating the appcast.xml to include the change information.  But where in the App is the best place to set the 'beta' channel.

I found this in the documentation for Sparkle

`func allowedChannels(for updater: SPUUpdater) -> Set<String> {
    return Set(["beta"])
}`

But i'm not sure where I would set it or call it.

Any help would be great.  
Thanks.

**Comment by zorgiepoo** [comment]
That method is a part of `SPUUpdaterDelegate`. You need to conform to `SPUUpdaterDelegate` protocol in one of your classes and implement that method. Then you have to set the updater's delegate to an instance of that class. If you are instantiating a `SPUStandardUpdaterController` from a nib, this is done from the `updaterDelegate` outlet. If you are creating an updater programmatically, this is done from one of the init methods.

In terms of the "best place", both the updater and its delegate should typically be around for your app's life cycle, so your NSApplication's delegate (if you're using Cocoa) could be one reasonable place. Of course it's up to you if you want to create more intermediate controllers and such if you programmatically create an updater.

**Reply by davidnorris78** [reply]
Thanks for getting back to me.  I'm using `SPUStandardUpdaterController` from a nib.  Is something like this okay?

`func applicationDidFinishLaunching(_ aNotification: Notification) {
        let updater = SPUStandardUpdaterController(updaterDelegate: self, userDriverDelegate: self).updater
        allowedChannels(for: updater)
  }`
`func allowedChannels(for updater: SPUUpdater) -> Set<String> {
        return Set(["beta"])
    }`

**Reply by zorgiepoo** [reply]
No. You don’t call the method directly. Sparkle calls it if you conform to the protocol. This is a basic use of the delegate pattern. And if you instantiate the updater in the nib you don’t create the updater or updater controller programmatically. You connect IBOutlets instead.

**Reply by davidnorris78** [reply]
i'm sorry, but it isn't very clear to me, with my level of knowledge and reading the documentation how exactly you link the two together.

I've got a menu button linked up in the nib to the `SPUStandardUpdaterController`. At the moment it's calling `checkForUpdates`.   How do I tell the `SPUStandardUpdaterController` to include the `beta` channel?  

Any help or other documentation you can recommend, would be a great help to me.

**Reply by zorgiepoo** [reply]
In your nib, make a connection from the updater controller (control click on it) updaterDelegate outlet to your app delegate https://imgur.com/a/4YVyoYb . Then you implement the allowedChannels method in your app delegate as you have, and Sparkle will ask it for the allowed channels when checking for updates.

Read up more on outlets, actions, and the delegation pattern if you don't have a good grasp because these are very fundamental Cocoa concepts.

**Reply by davidnorris78** [reply]
Thank you for your help.

**Reply by jackrlord** [reply]
Hi, im also trying to implement Beta channel and have gotten a little stuck, would you be able to point me in the right direction, I have the SPUUpdaterDelegate with allowedChannels returning ["Beta"]. However how do I get sparkle to point to a new url or sign a build into the Beta channel so it can be downloaded by the user? Sorry if im missing something obvious.

**Reply by zorgiepoo** [reply]
https://sparkle-project.org/documentation/publishing/#channels

You use the same feed but add the beta channel to the update item. If you’re using generate_appcast check its help page for how to add a channel.

**Reply by jackrlord** [reply]
Thank you! that was the link I was missing

**Reply by zorgiepoo** [reply]
You would have also found most of this information if you read the API documentation of the delegate method you were trying to implement..
